//@ compile-flags: --test

#![allow(soft_unstable)]
#![test]
<<<<<<< HEAD
//~^ ERROR 4:1: 4:9: `test` attribute cannot be used at crate level

// ferrocene-annotations: um_rustc_test
=======
//~^ ERROR `test` attribute cannot be used at crate level
>>>>>>> pull-upstream-temp--do-not-use-for-real-code
