//@ aux-build:hidden.rs

extern crate hidden;

use hidden::HiddenStruct;

struct InCrate {
    a: usize,
    b: bool,
    #[doc(hidden)]
    im_hidden: u8
}

fn main() {
    let HiddenStruct { one, two } = HiddenStruct::default();
    //~^ ERROR pattern requires `..` due to inaccessible fields

    let HiddenStruct { one } = HiddenStruct::default();
    //~^ ERROR pattern does not mention field `two` and inaccessible fields

    let HiddenStruct { one, hide } = HiddenStruct::default();
    //~^ ERROR pattern does not mention field `two`

    let InCrate { a, b } = InCrate { a: 0, b: false, im_hidden: 0 };
    //~^ ERROR pattern does not mention field `im_hidden`
}

// ferrocene-annotations: fls_gklst7joeo33
// External Crates
//
// ferrocene-annotations: fls_9gprp17h6t1q
// Use Imports
//
// ferrocene-annotations: fls_9ucqbbd0s2yo
// Struct Types
//
// ferrocene-annotations: fls_63v1fqedzwfd
// Attribute doc
//
// ferrocene-annotations: fls_7dbd5t2750ce
// Struct Patterns
//
// ferrocene-annotations: fls_nruvg0es3kx7
// Record Struct Patterns
//
// ferrocene-annotations: fls_9i5msiuuyihf
// Paths
//
// ferrocene-annotations: fls_151r19d7xbgz
// Entities
//
// ferrocene-annotations: fls_izl8iuhoz9e0
// Scopes
//
// ferrocene-annotations: fls_6ozthochxz1i
// Binding Scopes
//
// ferrocene-annotations: fls_ftphlagzd2te
// Generic Parameter Scope
//
// ferrocene-annotations: fls_m0z7omni9hp0
// Item Scope
//
// ferrocene-annotations: fls_769b4p8v3cwu
// Label Scope
//
// ferrocene-annotations: fls_kgbi26212eof
// Self Scope
//
// ferrocene-annotations: fls_octf6sf7yso
// Textual Macro Scope
//
// ferrocene-annotations: fls_lnpyb285qdiy
// Scope Hierarchy
//
// ferrocene-annotations: fls_dq403wq5yrs
// Namespaces
//
// ferrocene-annotations: fls_ld0ize96cm6m
// Preludes
//
// ferrocene-annotations: fls_ydmnb7qnmzzq
// Shadowing
//
// ferrocene-annotations: fls_40xoego2thsp
// Resolution
//
// ferrocene-annotations: fls_i6qzga6dyaee
// Path Resolution
//
// ferrocene-annotations: fls_o9u2h5m17kpz
// Path Expression Resolution
