<<<<<<< HEAD
#![cfg(FALSE)] //~ ERROR `main` function not found in crate `cfg_in_crate_1`

// ferrocene-annotations: fls_8jb3sjqamdpu
// Program Entry Point
=======
#![cfg(false)] //~ ERROR `main` function not found in crate `cfg_in_crate_1`
>>>>>>> pull-upstream-temp--do-not-use-for-real-code
